### PR TITLE
std: Minor changes to TLS handling

### PR DESCRIPTION
**Pull request by LemonBoy**
* Always allocate an info block per-thread so that libc can store
  important stuff there.
* Respect ABI-mandated alignment in more places.
* Nicer code, use slices/pointers instead of raw addresses whenever
  possible.

### CODE REVIEW

File: lib/std/os/linux/tls.zig [hunk 1]
```diff
@@ -195,124 +192,138 @@ pub fn initTLS() ?*elf.Phdr {
...
+        .tcb_offset = tcb_offset,
+        .dtv_offset = dtv_offset,
+        .data_offset = data_offset,
+        .gdt_entry_number = @bitCast(usize, @as(isize, -1)),
+    };
 }
 
-pub fn copyTLS(addr: usize) usize {
-    const tls_img = tls_image.?;
+inline fn alignPtrCast(comptime T: type, ptr: [*]u8) *T {
```
Comment (daurnimator) [inline lib/std/os/linux/tls.zig]: I think this function should be manually inlined.

Out of the two call-sites; one doesn't even need the `@alignCast`.

File: lib/std/os/linux/tls.zig [hunk 2]
```diff
@@ -195,124 +192,138 @@ pub fn initTLS() ?*elf.Phdr {
...
 
-    const slice = os.mmap(
-        null,
-        size,
-        os.PROT_READ | os.PROT_WRITE,
-        os.MAP_PRIVATE | os.MAP_ANONYMOUS,
-        -1,
-        0,
-    ) catch @panic("out of memory");
+    var tls_area = blk: {
```
Comment (daurnimator) [inline lib/std/os/linux/tls.zig]: s/var/const/

File: lib/std/thread.zig [hunk 3]
```diff
@@ -349,18 +348,21 @@ pub const Thread = struct {
                 else => return os.unexpectedErrno(@intCast(usize, err)),
             }
         } else if (std.Target.current.os.tag == .linux) {
-            var flags: u32 = os.CLONE_VM | os.CLONE_FS | os.CLONE_FILES | os.CLONE_SIGHAND |
-                os.CLONE_THREAD | os.CLONE_SYSVSEM | os.CLONE_PARENT_SETTID | os.CLONE_CHILD_CLEARTID |
-                os.CLONE_DETACHED;
-            var newtls: usize = undefined;
+            const flags: u32 = os.CLONE_VM | os.CLONE_FS | os.CLONE_FILES |
```
Comment (daurnimator) [inline lib/std/thread.zig]: Perhaps for ease of review going forward each flag should be on its own line?

File: lib/std/os/linux/tls.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/os/linux/tls.zig]: > I think this function should be manually inlined.

I made this small helper function as the inline casts produced too much line noise.

> Out of the two call-sites; one doesn't even need the @alignCast.

Both of them go from `*align(1) u8` to `*T` where the latter is ABI-aligned.

File: lib/std/os/linux/tls.zig (hunk 2 above)
Comment (LemonBoy) [reply lib/std/os/linux/tls.zig]: Good catch.

File: lib/std/thread.zig (hunk 3 above)
Comment (LemonBoy) [reply lib/std/thread.zig]: *shrug*, there are many flags and this set doesn't get modified that often.

File: lib/std/os/linux/tls.zig (hunk 1 above)
Comment (daurnimator) [reply lib/std/os/linux/tls.zig]: Line 292 is: `const tcb_ptr = alignPtrCast([*]u8, area.ptr + tls_image.tcb_offset);`
`[*]u8` should have an align of `1`, so `@ptrCast([*]u8, area.ptr + tls_image.tcb_offset)` should be sufficient.

Comment (LemonBoy) [reply lib/std/os/linux/tls.zig]: > [*]u8 should have an align of 1

It's not. The alignment is either 4 or 8.

File: lib/std/os/linux/tls.zig [hunk 4]
```diff
@@ -195,124 +192,138 @@ pub fn initTLS() ?*elf.Phdr {
...
+        // Fast path for the common case where the TLS data is really small,
+        // avoid an allocation and use our local buffer
+        if (tls_image.alloc_size < main_thread_tls_buffer.len) {
+            break :blk main_thread_tls_buffer[0..tls_image.alloc_size];
+        }
 
-    return @ptrToInt(slice.ptr);
+        break :blk os.mmap(
```
Comment (daurnimator) [inline lib/std/os/linux/tls.zig]: Should this have a check for `alloc_size > 0`? I don't think we want to do an `mmap` here if no tls is required.

File: lib/std/os/linux/tls.zig (hunk 1 above)
Comment (daurnimator) [reply lib/std/os/linux/tls.zig]: doh. I was thinking of the child; not the pointer itself.

File: lib/std/os/linux/tls.zig (hunk 4 above)
Comment (LemonBoy) [reply lib/std/os/linux/tls.zig]: The `alloc_size` is never zero.

> don't think we want to do an mmap here if no tls is required

That's the point, we want a per-thread local block of memory even if there's no threadlocal stuff.

**Comment by andrewrk** [comment]
Reverting for now since master branch is giving this:

```
576/903 behavior.misc.test "behavior-arm-linux-none-Debug-bare-multi thread local variable"...test failure
/home/vsts/work/1/s/lib/std/testing.zig:191:14: 0x4608f in std.testing.expect (test)
    if (!ok) @panic("test failure");
             ^
/home/vsts/work/1/s/test/stage1/behavior/misc.zig:616:11: 0x53e93 in behavior.misc.test "behavior-arm-linux-none-Debug-bare-multi thread local variable" (test)
    expect(S.t == 1235);
          ^
/home/vsts/work/1/s/lib/std/special/test_runner.zig:47:28: 0xcf73f in std.special.main (test)
        } else test_fn.func();
                           ^
/home/vsts/work/1/s/lib/std/start.zig:253:37: 0x7842b in std.start.posixCallMainAndExit (test)
            const result = root.main() catch |err| {
                                    ^
/home/vsts/work/1/s/lib/std/start.zig:123:5: 0x7820f in std.start._start (test)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
```

**Pull request merged**